### ISSUE TITLE
Auth/security suggestions

### ISSUE THREAD

**Issue by ferbs** [xshkut/roletalk-js#7]
I haven't yet decided on using a brokerless architecture for my current project but enjoyed poking around roletalk, thanks for sharing! 

I noticed a security-related issue along the way and wanted to pass it along. When comparing hmac signature to data.proof in Auth.ts#143 roletalk should use crypto.timingSafeEqual or a lib like tsscmp if you need to support older Node versions. ([Timing attack](https://codahale.com/a-lesson-in-timing-attacks/) vulnerability.)

**Comment by ferbs** [comment]
Also, consider moving the authentication logic to Peer#handleUpgrade and limiting the number of hmac signatures the recipient Peer will check. 

Since the initial ws connection looks like a normal http GET, external protective solutions like gateway throttling can be applied. And it simplifies the things, there's a single entrypoint with authentication required only once there, only authenticated peers can emit messages. 

During ws upgrade, you have access to the http request of the initial connection, allowing you to carry information embedded in the url path or searchParams and kindof in request headers. (If you need to support other ws clients like a browser, it would need to use a small hack, carrying header data on Sec-WebSocket-Protocol, with values passed in as "protocols".)

Re number of keys, currently an evil Peer could send vast numbers of fake signatures for a much lower cost than the receiving Peer must pay to verify. Better to have the user explicitly specify a small number of preshared keyIds to send when connecting, then only emit after authentication.

**Comment by xshkut** [comment]
> I haven't yet decided on using a brokerless architecture for my current project but enjoyed poking around roletalk, thanks for sharing! 
> 
> I noticed a security-related issue along the way and wanted to pass it along. When comparing hmac signature to data.proof in Auth.ts#143 roletalk should use crypto.timingSafeEqual or a lib like tsscmp if you need to support older Node versions. ([Timing attack](https://codahale.com/a-lesson-in-timing-attacks/) vulnerability.) 
> 

Hi ferbs,

Thanks for finding the vulnerability. I will replace the hash comparison method used with the one recommended.

**Comment by xshkut** [comment]
> Re number of keys, currently an evil Peer could send vast numbers of fake signatures for a much lower cost than the receiving Peer must pay to verify. Better to have the user explicitly specify a small number of preshared keyIds to send when connecting, then only emit after authentication. 

I agree, the number of keys sent should be limited and they should be explicitly specified on outgoing connection if registered keys exceed the limit.

**Comment by xshkut** [comment]
> Also, consider moving the authentication logic to Peer#handleUpgrade

Im afraid that will not be compatible with the existing challenge-response approach. 
Also, it is a common case when already have a server instance from which we need to retrieve an upgraded websocket and attach it to the roletalk instace.

**Comment by ferbs** [comment]
I'm imagining in-person events with strangers, like a gaming tournament or [Unconference](https://en.wikipedia.org/wiki/Unconference). Roletalk paired with some zero config service-discovery. (Not my current project, just like the idea.) 

If the hostile-peer scenario becomes relevant later, perhaps an additional token that is issued only on first upgrade and used alongside the existing challenge-response would help. I'm not a security expert and prob more would be required to address the scenario, but suspect it's on the right track.
